content/backend/graphql-scala/2-preparing-first-query.md

```diff
@@ -0,0 +1,377 @@
+    )
+)
+```
+
+</Instruction>
+
+We just added database definition of our first model.
+* **1** defines mapping to database table,
+* **2** gives us an helper we will use to accessing data in this table.
+* The last change, **3** is responsible for creating schema and adding three entities to the database.
+
+### Context
+
+Context is an object that flows across the whole execution, in the most cases this object doesn't change at all. The main responsibility of the Context it's to provide all the things that helps with fulfilling. In our example we will put there `DAO` so all queries will have access to the database, in the future we will also put there authentication data.
+
+<Instruction>
+
+Create `MyContext` class :
+
+```scala
+
+case class MyContext(dao: DAO)
+
+```
+
+</Instruction>
+
+### GraphQL Server
+
+Time to implement GraphQL Server. This Object will be in the second layer of architecture. Proper HTTP request will be converted into JSON object and send to this server. There, this object will be parsed, queries executed and sent response back to the client. It also will catch GraphQL parsing errors and converts it into the proper HTTP response.
+
+<Instruction>
+
+Create `GraphQLServer.scala` file:
+
+```scala
+
+import akka.http.scaladsl.server.Route
+import sangria.parser.QueryParser
+import spray.json.{JsObject, JsString, JsValue}
+import akka.http.scaladsl.model.StatusCodes._
+import akka.http.scaladsl.server.Directives._
+import scala.concurrent.ExecutionContext
+import scala.util.{Failure, Success}
+import akka.http.scaladsl.server._
+import sangria.ast.Document
+import sangria.execution._
+import akka.http.scaladsl.marshallers.sprayjson.SprayJsonSupport._
+import sangria.marshalling.sprayJson._
+
+
+object GraphQLServer {
+
+  // 1
+  private val dao = DBSchema.createDatabase
+
+  // 2
+  def endpoint(requestJSON: JsValue)(implicit ec: ExecutionContext): Route = {
+
+    // 3
+    val JsObject(fields) = requestJSON
+
+    // 4
+    val JsString(query) = fields("query")
+
+    // 5
+    QueryParser.parse(query) match {
+      case Success(queryAst) =>
+        // 6
+        val operation = fields.get("operationName") collect {
+          case JsString(op) => op
+        }
+
+        // 7
+        val variables = fields.get("variables") match {
+          case Some(obj: JsObject) => obj
+          case _ => JsObject.empty
+        }
+        // 8
+        complete(executeGraphQLQuery(queryAst, operation, variables))
+      case Failure(error) =>
+        complete(BadRequest, JsObject("error" -> JsString(error.getMessage)))
+    }
+
+  }
+
+  private def executeGraphQLQuery(query: Document, operation: Option[String], vars: JsObject)(implicit ec: ExecutionContext) = {
+    // 9
+    Executor.execute(
+      GraphQLSchema.SchemaDefinition, // 10
+      query, // 11
+      MyContext(dao), // 12
+      variables = vars, // 13
+      operationName = operation // 14
```

Comment:
If you got to 14 points in one snippet, then it's a sign that the complexity spiked in the paragraph :)